Some tools don't work without kernel-devel package when the kernel is build with ikheaders and BTF

### What reproduces the bug?

biosnoop.bt, dcsnoop.bt, mdflush.bt, runqlat.bt, tcpaccept.bt, tcpconnect.bt, rcpdrop.bt and tcpretrans.bt won't work when no kernel-devel package is installed, even though the kernel is build with CONFIG_IKHEADERS=m.
Ever since commit 17bbcd2b0773 ("kheader: Do not unpack kheaders if system has BTF"), we skip unpacking kheaders when BTF is available, but it seems those tools rely on data present in kheaders and not BTF. Reverting commit 17bbcd2b0773 fixes the issue.

Note that for some tools like biosnoop.bt, it just a matter of not finding the header file and removing the  #include line fixes the issue.
```
# /usr/share/bpftrace/tools/biosnoop.bt
definitions.h:2:10: fatal error: 'linux/blkdev.h' file not found
```

For other like dcsnoop.bt, it's complaining about incomplete type:
```
# /usr/share/bpftrace/tools/dcsnoop.bt 
definitions.h:5:25: error: field has incomplete type 'struct path'
definitions.h:5:16: note: forward declaration of 'struct path'
definitions.h:6:25: error: field has incomplete type 'struct qstr'
definitions.h:6:16: note: forward declaration of 'struct qstr'
```

### bpftrace --info
System
  OS: Linux 4.18.0-304.3.el8.bz1928463.x86_64 #1 SMP Wed Apr 14 11:46:12 EDT 2021
  Arch: x86_64

Build
  version: v0.12.1
  LLVM: 11.0.0
  ORC: v2
  foreach_sym: yes
  unsafe uprobe: no
  bfd: yes
  bpf_attach_kfunc: yes
  bcc_usdt_addsem: yes
  bcc bpf_attach_uprobe refcount: no
  libbpf: yes
  libbpf btf dump: yes
  libbpf btf dump type decl: yes

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: no

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf (depends on Build:libbpf): yes
  map batch (depends on Build:libbpf): yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): no

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  iter:task: yes
  iter:task_file: yes

--- a/docs/reference_guide.md
+++ b/docs/reference_guide.md
@@ -3522,7 +3522,8 @@ Histograms can also be printed on-demand, using the `print()` function. Eg:
 # BTF Support
 
 If kernel has BTF, kernel types are automatically available and there is no need to include additional headers
-to use them.
+to use them. To allow users to detect this situation in scripts, the preprocessor macro `BPFTRACE_HAVE_BTF` 
+is defined if BTF is detected. See tools/ for examples of its usage.
 
 Requirements for using BTF:
 

--- a/src/clang_parser.cpp
+++ b/src/clang_parser.cpp
@@ -748,6 +748,9 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
     // Since we're omitting <linux/types.h> there's no reason to
     // add the wokarounds for it
     args.push_back("-D__CLANG_WORKAROUNDS_H");
+    // Let script know we have BTF -- this is useful for prewritten tools to
+    // conditionally include headers if BTF isn't available.
+    args.push_back("-DBPFTRACE_HAVE_BTF");
 
     if (handler.parse_file("definitions.h", input, args, input_files, false) &&
         handler.has_redefinition_error())
@@ -778,6 +781,7 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
   {
     // There is a conflict (redefinition) between user-supplied types and types
     // taken from BTF. We cannot use BTF in such a case.
+    args.pop_back();
     args.pop_back();
     args.pop_back();
     input_files.back() = get_empty_btf_generated_header();

--- a/tools/biosnoop.bt
+++ b/tools/biosnoop.bt
@@ -1,5 +1,4 @@
 #!/usr/bin/env bpftrace
-#include <linux/blkdev.h>
 /*
  * biosnoop.bt   Block I/O tracing tool, showing per I/O latency.
  *               For Linux, uses bpftrace, eBPF.
@@ -10,6 +9,10 @@
  *
  * 15-Nov-2017	Brendan Gregg	Created this.
  */
+
+#ifndef BPFTRACE_HAVE_BTF
+#include <linux/blkdev.h>
+#endif
 
 BEGIN
 {

--- a/tools/dcsnoop.bt
+++ b/tools/dcsnoop.bt
@@ -15,6 +15,7 @@
  * 08-Sep-2018	Brendan Gregg	Created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/fs.h>
 #include <linux/sched.h>
 
@@ -24,6 +25,7 @@ struct nameidata {
         struct qstr     last;
         // [...]
 };
+#endif
 
 BEGIN
 {

--- a/tools/mdflush.bt
+++ b/tools/mdflush.bt
@@ -13,8 +13,10 @@
  * 08-Sep-2018	Brendan Gregg	Created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/genhd.h>
 #include <linux/bio.h>
+#endif
 
 BEGIN
 {

--- a/tools/naptime.bt
+++ b/tools/naptime.bt
@@ -13,8 +13,10 @@
  * 16-Feb-2019  Brendan Gregg   Created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/time.h>
 #include <linux/sched.h>
+#endif
 
 BEGIN
 {

--- a/tools/oomkill.bt
+++ b/tools/oomkill.bt
@@ -20,7 +20,9 @@
  * 07-Sep-2018	Brendan Gregg	Created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/oom.h>
+#endif
 
 BEGIN
 {

--- a/tools/runqlen.bt
+++ b/tools/runqlen.bt
@@ -11,17 +11,19 @@
  * 07-Oct-2018	Brendan Gregg	Created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/sched.h>
 
 // Until BTF is available, we'll need to declare some of this struct manually,
 // since it isn't available to be #included. This will need maintenance to match
 // your kernel version. It is from kernel/sched/sched.h:
-struct cfs_rq_partial {
+struct cfs_rq {
 	struct load_weight load;
 	unsigned long runnable_weight;
 	unsigned int nr_running;
 	unsigned int h_nr_running;
 };
+#endif
 
 BEGIN
 {
@@ -31,7 +33,7 @@ BEGIN
 profile:hz:99
 {
 	$task = (struct task_struct *)curtask;
-	$my_q = (struct cfs_rq_partial *)$task->se.cfs_rq;
+	$my_q = (struct cfs_rq *)$task->se.cfs_rq;
 	$len = $my_q->nr_running;
 	$len = $len > 0 ? $len - 1 : 0;	// subtract currently running task
 	@runqlen = lhist($len, 0, 100, 1);

--- a/tools/tcpaccept.bt
+++ b/tools/tcpaccept.bt
@@ -16,8 +16,12 @@
  * 23-Nov-2018	Dale Hamel	created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/socket.h>
 #include <net/sock.h>
+#else
+#include <sys/socket.h>
+#endif
 
 BEGIN
 {

--- a/tools/tcpconnect.bt
+++ b/tools/tcpconnect.bt
@@ -19,8 +19,12 @@
  * 23-Nov-2018	Dale Hamel	created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/socket.h>
 #include <net/sock.h>
+#else
+#include <sys/socket.h>
+#endif
 
 BEGIN
 {

--- a/tools/tcpdrop.bt
+++ b/tools/tcpdrop.bt
@@ -17,8 +17,12 @@
  * 23-Nov-2018	Dale Hamel	created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/socket.h>
 #include <net/sock.h>
+#else
+#include <sys/socket.h>
+#endif
 
 BEGIN
 {

--- a/tools/tcplife.bt
+++ b/tools/tcplife.bt
@@ -13,10 +13,14 @@
  * 17-Apr-2019  Brendan Gregg   Created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <net/tcp_states.h>
 #include <net/sock.h>
 #include <linux/socket.h>
 #include <linux/tcp.h>
+#else
+#include <sys/socket.h>
+#endif
 
 BEGIN
 {

--- a/tools/tcpretrans.bt
+++ b/tools/tcpretrans.bt
@@ -17,8 +17,12 @@
  * 23-Nov-2018  Dale Hamel      created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <linux/socket.h>
 #include <net/sock.h>
+#else
+#include <sys/socket.h>
+#endif
 
 BEGIN
 {

--- a/tools/tcpsynbl.bt
+++ b/tools/tcpsynbl.bt
@@ -13,7 +13,9 @@
  * 19-Apr-2019  Brendan Gregg   Created this.
  */
 
+#ifndef BPFTRACE_HAVE_BTF
 #include <net/sock.h>
+#endif
 
 BEGIN
 {

